deps/build.jl

```diff
@@ -24,17 +24,9 @@ if install_gap
     println("Installing GAP ...")
     gap_branch = "stable-4.11"
     cd(extra_gap_root)
-    ## check if gap already exists
-    if isdir("gap")
-        ### Update the GAP clone
-        cd("gap")
-        run(`git fetch origin`)
-        run(`git checkout $(gap_branch)`)
-        run(`git pull origin $(gap_branch)`)
-    else
-        run(`git clone -b $(gap_branch) https://github.com/gap-system/gap`)
-        cd("gap")
-    end
+    run(`rm -rf gap`)
```

Comment:
I'm no git guru ;) but maybe this will be better:
```julia
try
    if isdir("gap")
        cd("gap")
        run(`git checkout $(gap_branch)`)
        run(`git pull origin $(gap_branch)`)
    else
        mkdir("gap")
        cd("gap")
        run(`git remote add origin https://github.com/gap-system/gap`)
        run(`git fetch --depth 100 origin $(gap_branch))
        run(`git checkout $(gap_branch)`)
    end
catch e
    @error("Error occured while updating gap repository at $(pwd())
    rethrow(e)
finally
    cd(orig_dir)
end
```
in general whenever I need to `cd` I try to `try-catch-finally`

Reply:
No that won't work for multiple reasons
- `gap` may be an existing dir but not a git repository (indeed in the future, this is intended to be the default; it already would be, if the GAP 4.11 release hadn't been delayed so badly)
- even if it is a `.git` repository, it'll be a shallow clone; while un-shallowing supposedly works pretty well these days, I am wary of it due to issues I had with it in the past
- even if it is a `.git` repository, it's not guaranteed that `origin` is right (but that could be fixed by specifying the remote URL explicitly, I guess)
- in the `else` branch, that code won't work at all (to what repository are you adding that `origin` remote? Well, to the enclosing git repository... which is not the empty `gap` dir you just created, but rather the dir of the Julia package; which probably already has a remote `origin`, so it'll just run into an error). The fix is to use `git clone --depth=...` -- which is precisely what my PR does :-)

Good point about that try/catch, I'll add that.

Reply:
Hmm, I guess I'd best wrap all of `build.jl` into a try/catch then... but that makes me wonder: why doesn't `Pkg.jl` do that (save & restore PWD) for you? Or perhaps it does (hmm, nah, at least on a quick look through its code base, it doesn't seem to; ah well)

Reply:
I think I forgot `git init`, but looking at the documentation of `git clone --depth` I guess it does in one step what I did in three ;-)
```julia
try
           if isdir("gap")
               cd("gap")
               run(`git checkout $(gap_branch)`)
               run(`git pull origin $(gap_branch)`)
           else
               mkdir("gap")
               cd("gap")
               run(`git clone --depth 100 -b $(gap_branch) https://github.com/gap-system/gap` )
           end
       catch e
           @error("Error occured while updating gap repository at $(pwd())")
           rethrow(e)
       finally
           cd(orig_dir)
       end
```
otherwise it works (just checked);
 * I guess the future is BinaryGAP, so this becomes a no-issue
 * there is no need to unshallow it
 * if it is a user supplied directory with git-tree it's also user responsibility to point origin to the right remote
 * all errors are handled gracefully and the user is informed what happened, e.g. without `git init`:
```
error: pathspec 'stable-4.11' did not match any file(s) known to git
┌ Error: Error occured while updating gap repository at /home/kalmar/.julia/dev/GAP/gap
└ @ Main REPL[14]:15
ERROR: failed process: Process(`git checkout stable-4.11`, ProcessExited(1)) [1]
```

Reply:
Perhaps it can be made to work, but what's the usecase? On the downside, it means people can end up in weird unsupported states. If you are requesting us to build gap for you, let us do it, by starting from a clean, known-to-be-good slate; if you want a custom GAP, just don't ask us to build it for you, and supply your own.

Reply:
the usecase is that we don't download full git tree (accomplished) and gap tarball (still downloaded) for packages every time `build` is run. But I guess we need to live with this for the time